[Messenger] Add a scheduler component

upyx opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | no
| License       | MIT
| Doc PR        | TBD


### Introdution

There is no easy way to schedule periodical tasks. There are few useful tools which I touched:

- https://github.com/Guikingone/SchedulerBundle
- https://github.com/Cron/Symfony-Bundle
- https://github.com/zenstruck/schedule-bundle
- https://github.com/lavary/crunz

They are complicated. They doesn't allow to set time with precision in seconds (at least it isn't easy). They require difficult tricks to set not linear periods (like on sunset in Tokyo). ~They are~ Part of them inefficient with infrequent tasks because resources are needed on every run (e.g. Kubernetes [CronJob](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/) creates and destroes an entire container).

### Proposal

I use a custom transport of Messenger that **generates** messages periodically. It's a simple solution to run periodical jobs without external software. It's especially helpful in environments where additional runners are impossible or very painful (like Kubernetes). Configuration is very flexible and accurate, it's possible to configure any rules when to run or not to run.

Compared to `crond`: there is no need to install something external, precision of schedule in microseconds instead of *minutes*, it's possible to set any periods.

#### Simple example

```yaml
# messenger.yaml
framework:
  messenger:
    transports:
      schedule_default: 'schedule://default'
```

Few types of messages:
```php
class SomeJob {}
class OtherJob {}
```

A handlers:
```php
#[AsMessageHandler]
class SomeJobHandler {
    public function __invoke(SomeJob $job) {
        // do job or delegate it to other service
    }
}

#[AsMessageHandler]
class OtherJobHandler {
    public function __invoke(OtherJob $job) {
        // do job or delegate it to other service
    }
}
```

A schedules are provided by locators. It's possible to create many locators and/or provide many schedules by the same locator:

```php
class ExampleLocator implements ScheduleLocatorInterface
{
    public function get(string $id): ScheduleConfig
    {
        // once after an hour from now
        $deferForAnHour = new \DatePeriod(
            new \DateTimeImmutable('now'),
            new \DateInterval('PT1H'),
            1,
            \DatePeriod::EXCLUDE_START_DATE
        );

        return (new ScheduleConfig())
            ->add(new OnceTrigger(new \DateTimeImmutable()), new WarmUpJob()) // once on every worker's start
            ->add(PeriodicalTrigger::create('P1D', '2022-01-01T03:00:00Z'), new SomeJob()) // every night at 3 a.m. (UTC)
            ->add(PeriodicalTrigger::fromPeriod($deferForAnHour), new OtherJob())
        ;
    }

    public function has(string $id): bool
    {
        return 'default' === $id;
    }
}
```

To run schedule:

```sh
bin/console messenger:consume schedule_default
```

It's easy to run jobs manually:

```php
#[AsCommand(name: 'some', description: 'Manually runs SomeJob')]
class SomeCommand extends Command {
    public function __construct(private MessageBusInterface $bus) {
    }

    protected function execute(InputInterface $input, OutputInterface $output): int {
        $this->bus->dispatch(new SomeJob());
    }
}
```

#### Example with returning a result

```php
class GoodJob {
    public function __construct(public readonly ?LoggerInterface $logger) {
    }
}
```

```php
#[AsMessageHandler]
class GoodHandler {
    public function __construct(private readonly LoggerInterface $logger) {
    }

    public function __invoke(GoodJob $job){
        // compute $result
        ($job->logger ?? $this->logger)->info('The result is: {result}', ['result' => $result])
    }
}
```

```php
#[AsCommand(name: 'job', description: 'Manually runs job')]
class SomeCommand extends Command {
    public function __construct(private MessageBusInterface $bus) {
    }

    protected function execute(InputInterface $input, OutputInterface $output): int {
        $this->bus->dispatch(new GoodJob(new ConsoleLogger($output))); // result will be printed in console
    }
}
```

#### Configuring

A minimal configuration:

```yaml
# messenger.yaml
framework:
  messenger:
    transports:
      schedule_default: 'schedule://default'
```

More complex example:

```yaml
# messenger.yaml
framework:
  messenger:
    transports:
      schedule_default:
        dsn: 'schedule://default'
        options:
          cache: 'app.cache'
          lock: 'default'
```

Example HA configuration with redis:

```yaml
framework:
  cache:
    default_redis_provider: '%env(REDIS_DSN)%'
  lock:
    redis: '%env(REDIS_DSN)%'
  messenger:
    transports:
      schedule_default:
        dsn: 'schedule://default'
        options:
          cache: 'cache.redis'
          lock:
            resource: 'redis'
            ttl: 60
            auto_release: true
```

### Deprecations

None

### Implementation

This PR contains an implementation.

### ToDo

- [x] Remove obsolete code
- [x] Add a configuration to the Framework
- [x] Specialize exceptions and improve messages
- [x] Cover with tests
- [ ] Add acceptance tests for HA 
- [x] Fix `CHANGELOG`s & `README`s
- [ ] Add documentation

upyx committed:
Add the Scheduler component

Guikingone commented:
Hi 👋🏻 

As `SchedulerBundle` and the related PR has been linked, some informations about the related "solutions" and some assertions that you made: 

> The existing solutions (https://github.com/symfony/symfony/issues/39719, [2](https://github.com/Cron/Symfony-Bundle), more?) depend on additional software like crond. 

Regarding `SchedulerBundle`, it DOES NOT rely on `crond`, `crond` is A way to launch / consume tasks but the main logic behind SchedulerBundle is to be able to consume tasks no matter how you need to, it could be an HTTP call, using the command or just by calling the worker.

> They doesn't allow to set time with precision in seconds.

Actually, it depends, talking about `SchedulerBundle`, you can set an expression via a custom `ExpressionFactory`, if precision like `MonotonicClock` is required, it could be extended when needed, the interface is autoconfigured.

> They require difficult tricks to set not linear periods (like on sunset in Tokyo). They are inefficient with frequent tasks because resources are needed on every run (e.g. Kubernetes [CronJob](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/) runs an entire container).

In the end, running the task via the worker of `Messenger` or via a dedicated worker is quite the same, a ressource is always consumed when you consume a task 🙂

upyx commented:
Nice to meet you, @Guikingone :wave:

I've mentioned `SchedulerBundle` because it's an awesome tool for cron like schedules, and it's a good for requests&tasks based applications (which are the most widespread... perhaps :slightly_smiling_face: )

> Regarding SchedulerBundle (...), it DOES NOT rely on crond, crond is A way to launch / consume tasks ...

But it **[requires](https://github.com/zenstruck/schedule-bundle/blob/master/doc/run-schedule.md) a way** to launch tasks! Without running `ScheduleRunner::__invoke()` it won't work. This PR introduces yet another way to do that with Messenger. It's absolutely legal to use them together :slightly_smiling_face:

e.g.

```php
class ScheduleFactory {
    public static function createSchedule(ClockInterface $clock){
        return new \Symfony\Component\Messenger\Bridge\Scheduler\Transport\Schedule(
            $clock,
            PeriodicalJob::infinite('00:00:00', 'PT1M', new EveryMinuteJob()),
        );
    }
}
```

```php
#[AsMessageHandler]
class EveryMinuteHandler {
    public function __construct(private readonly ScheduleRunner $scheduleRunner) {
    }

    public function __invoke(EveryMinuteJob $job){
        ($this->scheduleRunner)();
    }
}
```

> In the end, running the task via the worker of Messenger or via a dedicated worker is quite the same, a ressource is always consumed when you consume a task :slightly_smiling_face:

Yes, it's possible to write some custom worker with `while` loop. But it isn't necessary. The `Messenger`'s `Worker` does a loop, and does it very well (with count and memory limits and other nice features).

upyx commented:
Nevertheless, there is a **nuance**... Hm... there are few ones :grin: 

Why is some Schedule/Cron/Jobs/Tasks/Etc bundle is needed when the crond/cronspec/CronJob/cronetc is already exists? Why do not configure a crontab/CloudSpec/Helm/Etc to run tasks?

The answer is: because it is necessary to control the application's tasks in the application itself, not by infrastructure.

The first nuance is the Messenger is in the application already (no needs to launch anything dependent on time).  
The second one is the `ScheduleTransport` does the most of the ALotOfBundles features by less than 500 lines of code.

ArcaneDrift commented:
I like this idea, been able to keep the application code together, having all "Schedule/Queue" tasks together helps with maintainability and without needing crond would be a very nice touch to have in some container setups.

fabpot commented:
@upyx I like it a lot :)

upyx commented:
@Guikingone I was cofused with zenstruck's bundle and yours, sorry for that. A broken link in the issue (#39719) supports doing mistakes like that.
[Guikingone/SchedulerBundle](https://github.com/Guikingone/SchedulerBundle) has the `--wait` option on the `scheduler:consume` command that can be used without any `crond` or so on. However, the other values are still valid.

@fabpot I'll prepare PR for review in few days.

Guikingone commented:
> @Guikingone I was cofused with zenstruck's bundle and yours, sorry for that. A broken link in the issue (https://github.com/symfony/symfony/issues/39719) supports doing mistakes like that.
[Guikingone/SchedulerBundle](https://github.com/Guikingone/SchedulerBundle) has the --wait option on the scheduler:consume command that can be used without any crond or so on. However, the other values are still valid.

No problem 😅 

I tend to have a pretty different vision of why a scheduler worker is different from a message queue one (in particular when it comes to consuming tasks, handling failure, transports, limits, middleware and so on), in the end, the way the linux scheduler handles tasks is not the same as RabbitMQ or even Kafka but that's a complete separate topic from the current one 🙂

upyx commented:
Ready for feedback :writing_hand: 

Few questions:

 - The Bridge directory is not the best place for the component. Where to place it?
 - Where to place the `testMessengerScheduler`?
 - I can't choose how to implement configuration of schedules: interfaces, attributes, factories?
 - Should I "fix" psalm warnings or can suppress them?
 - Is it possible to teach fabbot the word "kaz-info-teh"? :rofl:

upyx commented:
@Guikingone

> I tend to have a pretty different vision of why a scheduler worker is different from a message queue one ...

I understand what you are saying about. Despite much in common, message processing and task scheduling are completely different things. It matters when we design an application that deals with realtime signals like a shooter game or an audio editor. It matters when we plan a distributed application like Telegram. When the only thing that is needed is to create a new table partition every week, I tend to save a few key presses and some space in the vendor directory.

fabpot commented:
@upyx Bridges are only helpful when a feature depends on a third-party service/prodiver. Here, that's not the case, so I would move the code to the "core" of the component instead.

fabpot commented on src/Symfony/Component/Messenger/Transport/Schedule/JobScheduleInterface.php:
```diff
@@ -0,0 +1,19 @@
+
+namespace Symfony\Component\Messenger\Transport\Schedule;
+
+interface JobScheduleInterface
```
As jobs are messages (as we call them in Messenger), what about renaming this interface to `ScheduledMessageInterface`?

upyx replied:
Well, it is not a message itself. It's a schedule that provides messages (by `getJob`) on time when they are needed to be done. ~Maybe `ScheduledTaskInterface` and `getMessage()` method? Then `PeriodicalTask` which implements `ScheduledTaskInterface`~
 
The `ScheduledTaskInterface` sounds even worse. I have no idea...

upyx replied:
I used "job" over `task` because the word "task" is heavily overused.

upyx replied:
Well, I have renamed it as you advised :)

fabpot commented on src/Symfony/Component/Messenger/Transport/Schedule/PeriodicalJob.php:
```diff
@@ -0,0 +1,95 @@
+
+namespace Symfony\Component\Messenger\Transport\Schedule;
+
+final class PeriodicalJob implements JobScheduleInterface
```
PeriodicalMessage?

fabpot commented on src/Symfony/Component/Messenger/Transport/Schedule/PeriodicalJob.php:
```diff
@@ -0,0 +1,95 @@
+            if ('P' === $interval[0]) {
+                $interval = new \DateInterval($interval);
+            } else {
+                \assert((string) (int) $interval === $interval);
```
Is it really useful?

upyx replied:
Oh... I will change it to `ctype_digit`, it would be more clear. It's a check that string contains integers only.

fabpot reviewed (changes requested):
I still need to play with it a bit more, but I like the simplicity of this approach.
I've made some renaming suggestions.
~~Would it sense to deprecate `DelayStamp` as it does something similar?~~

fabpot commented on src/Symfony/Component/Messenger/Transport/Schedule/PeriodicalMessage.php:
```diff
@@ -0,0 +1,102 @@
+        $startDate = \DateTimeImmutable::createFromInterface($period->getStartDate());
+        $nextDate = $startDate->add($period->getDateInterval());
+        $interval = $nextDate->getTimestamp() - $startDate->getTimestamp();
+        $fistRun = $period->include_start_date ? $startDate : $nextDate;
```
```suggestion
        $firstRun = $period->include_start_date ? $startDate : $nextDate;
```

fabpot commented on src/Symfony/Component/Messenger/Transport/Schedule/PeriodicalMessage.php:
```diff
@@ -0,0 +1,102 @@
+                    ->modify($delta % 86400 .'seconds');
+        }
+
+        return new self($message, $interval, $fistRun, $priorTo);
```
```suggestion
        return new self($message, $interval, $firstRun, $priorTo);
```

fabpot commented on src/Symfony/Component/Messenger/Transport/Schedule/ScheduleTransportFactory.php:
```diff
@@ -0,0 +1,44 @@
+            throw new InvalidArgumentException(sprintf('The given Schedule DSN "%s" is invalid.', $dsn));
+        }
+
+        $scheduleName = $options['name'] ?? $parsedUrl['host'] ?? 'default';
```
```suggestion
        $scheduleName = $parsedUrl['host'] ?? 'default';
```

I don't see why we would support both.

upyx replied:
All bridges do it :man_shrugging: 

Tipically, a DSN is got from env vars, and it can partially overwrite defaults from options. It's very useful when an application is distributed to many nodes or/and env vars are changed in runtime.

fabpot replied:
But here, we are configuring a "static" DSN where the host is a static/random name. I don't think this will ever be configured via a DSN and it will never be overridden. Or am I missing something?

upyx replied:
I agree with you :+1: Let's simplify it.

upyx commented:
I have used the new Clock component. How to properly add a dependency of that? `"symfony/clock": "^6.2"`?

fabpot commented on src/Symfony/Component/Messenger/Transport/Schedule/Schedule.php:
```diff
@@ -0,0 +1,70 @@
+ */
+const _SCHEDULE_FAR_FUTURE_ = new \DateTimeImmutable('3000-01-01');
+
+class Schedule implements \IteratorAggregate
```
```suggestion
final class Schedule implements \IteratorAggregate
```

upyx replied:
Are you sure?
I did it exactly as you suggest. And have changed it after playing with it.

The easiest way to provide a schedule it extend the `Schedule` class, ond overwrite the constructor:

```php
#[AsSchedule('default')]
class MySchedule extends Schedule {
    public function __construct(ClockInterface $clock) {
        parent::__construct(
            $clone,
            new PeriodicalMessage(/*...*/),
            /* others ... */
        );
    }
}
```

That's it! No factories, no implementations of interfaces, no specific configurations...

And it's optional. It's possible to configure a service in any way.

fabpot replied:
Indeed, forget about this one.

fabpot commented on src/Symfony/Component/Messenger/Transport/Schedule/ScheduleTransport.php:
```diff
@@ -0,0 +1,49 @@
+
+class ScheduleTransport implements TransportInterface
+{
+    public function __construct(private readonly \Traversable $schedule)
```
Let's create an interface instead:

```
interface ScheduleInterface
{
    public function getSchedule(): iterable;
}
```

upyx replied:
I did it :rofl: I can return it, if you wish.

The idea, why I removed it is: the message it's any object of any class, and the schedule it's an iterable of that objects. So, it's possible to do any iterable and pass it to transport to provide messages. And it will work.

fabpot replied:
I still prefer to have an interface.

fabpot commented on src/Symfony/Component/Messenger/composer.json:
```diff
@@ -22,6 +22,7 @@
     "require-dev": {
         "psr/cache": "^1.0|^2.0|^3.0",
         "symfony/console": "^5.4|^6.0",
+        "symfony/clock": "^6.2",
```
Should be before console

upyx replied:
Oh, by bad.
Hovewer it doesn't help with tests: "Error: Class "Symfony\Component\Clock\MockClock" not found"

lyrixx commented:
I didn't fully read the code but I have some concerns if I understand correctly

So there is a message generator that are triggered when wanted 

So what does happen when the worker is down while a message should have been dispatched ?

And what does happen when you have N server where the worker is installed? N messages are sent?

fabpot commented:
> I didn't fully read the code but I have some concerns if I understand correctly
> 
> So there is a message generator that are triggered when wanted
> 
> So what does happen when the worker is down while a message should have been dispatched ?
> 
> And what does happen when you have N server where the worker is installed? N messages are sent?

I have the same concerns and then some more. All of them would be the same with a new Scheduler/Cron/... component. They do exist and we will address them.

Some thoughts about your 2 concerns:

 * If the server is down, the messages cannot be dispatched, but the same goes for crons (or any other scheduler that you can use that runs on a server).
 * For the N servers issue, we can use locks (and then, you don't have the down issue anymore).

lyrixx commented:
It would be nice to have built in lock support to get HA out of the box. I imagine a simple configuration option where you put the lock provider id.

fabpot commented:
I'm preparing some changes on top of this PR to make it possible to add more features (like locks, multi-server support, ...).

lyrixx commented:
👍🏼 awesome

I like this feature!

upyx commented:
> So what does happen when the worker is down while a message should have been dispatched ?

The message won't be dispatched in that case. At first, I did a cache for `$lastTick` to dispatch "skipped" messages, but then removed it. It more close to cron-like logic.

> 
> And what does happen when you have N server where the worker is installed? N messages are sent?

Yes, N messaages are sent. They works separatelly.

upyx commented:
> I'm preparing some changes on top of this PR to make it possible to add more features (like locks, multi-server support, ...).

OMG I'm too :) I have hoped to do it bit by bit.

fabpot commented:
> > I'm preparing some changes on top of this PR to make it possible to add more features (like locks, multi-server support, ...).
> 
> OMG I'm too :) I have hoped to do it bit by bit.

Ok, I will stop for now. Here are 2 commits that I think should be included: https://github.com/upyx/symfony/compare/messenger-scheduler-transport...fabpot:symfony:messenger-scheduler-transport?expand=1

The first one allows controlling if the message should be handled immediately or resent to the bus so that it can be handled like any other message. That allows having one consumer that only does the scheduling and is not slowed down by the actual message handling (which can be done by another consumer).

The second commit introduces a TriggerInterface and kills the `ScheduledMessageInterface` interface in favor of a `ScheduledMessage`. That gives more flexibility on configuration (think locks for instance).

upyx commented:
@fabpot 
Would you like to make one PR with all the things?

fabpot commented:
> @fabpot Would you like to make one PR with all the things?

Feel free to steal my code :)

fabpot commented:
@upyx Are you still working on this? No pressure of course but I'd like to get this feature in 6.2 if possible. For your information, code freeze is 3/4 weeks away, and as we will need quite a few rounds of reviews, we should probably move forward in the coming couple of weeks. Tell me if you need help.

upyx commented:
@fabpot Yes, I'm in progress. I'll do my best.

jderusse commented on src/Symfony/Component/Messenger/Transport/Schedule/Schedule.php:
```diff
@@ -0,0 +1,70 @@
+        }
+
+        $nearestNextRun = self::FAR_FUTURE;
+        foreach ($this->scheduledMessages as $scheduledMessage) {
```
the implementation is not optimal,
- schedulers are called twice (once for the next run in the future, and once again when lastTick<nextRun<now)
- all schedulers are checked every time one of then is run (if you have one scheduler for `PT1S` and 100 schedulers for `P1D` the 100 schedulers are checked everyv second even if they run once a day)

what's about initializing a table of next runs, then using that table to check if messages should be yield. When a message is yield then add the next run in the table.

upyx replied:
It's interesting that you've found it :+1:

> schedulers are called twice

On the first run only, I have fixed it.

> all schedulers are checked every time one of then is run

Yes. It significant simplifies an implementation.

> what's about initializing a table of next runs

I have thought about it. I believe it is not necessary. It's a common dilemma: to remember or to compute. If a call `getNextRun()` would be expensive it can be cached (internally), the `PeriodicalTrigger` is simple enough to be called every time.

I'm preparing the next version of implementation. It will work differently and will have different issues :)

upyx replied:
I have used the `SplHeap` as a "table". Looks good. Could you check?

jderusse commented on src/Symfony/Component/Messenger/Transport/Schedule/Schedule.php:
```diff
@@ -0,0 +1,70 @@
+
+class Schedule implements \IteratorAggregate
+{
+    public const FAR_FUTURE = _SCHEDULE_FAR_FUTURE_;
```
this could be private: this variable is publicly used in one case: when the next run is after the end of the period. in that case the name is wrong : it's not "far in the future" but should be "never". 
And the scheduler could return null to tell the transport "don't call me again".

upyx replied:
Personally, I hate returning nulls as special case. But that constant seems more confusing, so I removed it.

upyx commented:
@lyrixx 
> It would be nice to have built in lock support to get HA out of the box. I imagine a simple configuration option where you put the lock provider id.

Well. I did something like that. 

```yaml
framework:
  cache:
    default_redis_provider: '%env(REDIS_DSN)%'
  lock:
    redis: '%env(REDIS_DSN)%'
  messenger:
    transports:
      cron:
        dsn: 'schedule://default'
        options:
          cache: 'cache.redis'
          lock: 'redis'
```

Is that what you meant?

upyx commented:
@fabpot it's ready for feedback. I hope it will be merged to 6.3.

fabpot commented on src/Symfony/Component/Messenger/Transport/Schedule/Progress/ProgressFactory.php:
```diff
@@ -0,0 +1,79 @@
+{
+    public function __construct(
+        private readonly ClockInterface $clock,
+        private readonly ContainerInterface $lockFactories,
```
We should probably accept `null` here.

fabpot commented on src/Symfony/Component/Messenger/Transport/Schedule/Progress/ProgressFactory.php:
```diff
@@ -0,0 +1,79 @@
+    public function __construct(
+        private readonly ClockInterface $clock,
+        private readonly ContainerInterface $lockFactories,
+        private readonly ContainerInterface $caches,
```
Same here

fabpot commented on src/Symfony/Component/Messenger/Transport/Schedule/Progress/ProgressFactory.php:
```diff
@@ -0,0 +1,79 @@
+
+    private function createLock(string $scheduleName, string $resourceName, array $options): LockInterface
+    {
+        if (!($options['lock'] ?? false)) {
```
If `symfony/lock` is not installed, we should throw an exception (as lock is not part of the required packages for Messenger).

fabpot reviewed (commented):
I have tested quickly this new implementation and I like it a lot. I still need to play with it more, but I think we are heading in the right direction. Thank you so much for your commitment.

Now that the implementation is composed of many classes, I'm wondering if we should create a new Scheduler component that would depend on Messenger? We could have shorter namespaces, Lock/Clock/... as "real" dependencies, and better discoverability. That would also avoid "polluting" the Transport sub-namespace as most classes are not about the transport itself. WDYT?

fabpot commented on src/Symfony/Component/Messenger/Transport/Schedule/Locator/ChainScheduleLocator.php:
```diff
@@ -0,0 +1,64 @@
+    /**
+     * @param iterable<ScheduleLocatorInterface> $locators
+     */
+    public function __construct(iterable $locators)
```
```suggestion
    public function __construct(
        private iterable $locators,
    )
```

upyx replied:
The `iterable` (actually "tagged_iterator") is converted to an array and types are checked by trick with a closure. It's possible to skip type checking, but converting an iterator to an array is still needed.

fabpot commented on src/Symfony/Component/Messenger/Transport/Schedule/Schedule.php:
```diff
@@ -0,0 +1,106 @@
+    private ProgressInterface $progress;
+    private ScheduleHeap $scheduleHeap;
+
+    public function __construct(ScheduleConfig $scheduleConfig, ProgressInterface $progress)
```
```suggestion
    public function __construct(
        ScheduleConfig $scheduleConfig,
        private ProgressInterface $progress,
    )
```

upyx replied:
Ok... But IMO, mixing promoted properties and regular ones is ugly.

fabpot commented on src/Symfony/Component/Messenger/Transport/Schedule/ScheduleConfig.php:
```diff
@@ -0,0 +1,47 @@
+    /**
+     * @param iterable<array{TriggerInterface, object}> $schedule
+     */
+    public function __construct(iterable $schedule = [])
```
```suggestion
    public function __construct(
        private iterable $schedule = [],
    )
```

upyx replied:
It won't work like that. The "itarable" is iterated and add() method is called for types checking.

stof replied:
this iteration in the constructor will break lazy-loading. Please remove it.

fabpot commented on src/Symfony/Component/Messenger/Transport/Schedule/ScheduleTransport.php:
```diff
@@ -0,0 +1,52 @@
+    private ScheduleInterface $schedule;
+    private array $stamps;
+
+    public function __construct(ScheduleInterface $scheduler)
```
```suggestion
    public function __construct(
        private ScheduleInterface $schedule,
    )
```

I've also renamed scheduler to schedule.

fabpot commented on src/Symfony/Component/Messenger/Transport/Schedule/Schedule.php:
```diff
@@ -0,0 +1,106 @@
+    }
+}
+
+/**
```
Should be moved to its own file

fabpot commented on src/Symfony/Component/Messenger/Transport/Schedule/Progress/Progress.php:
```diff
@@ -0,0 +1,90 @@
+    {
+        $this->nextTime = $nextTime;
+
+        unset($this->time, $this->index);
```
We should probably set them to `null` instead.

upyx replied:
Probably...

The `time` and the `index` properties are not nullable, but must be undefined before `acquire` and after `release`.

upyx replied:
I've added exceptions. I hope they help to make logic clear.

fabpot commented on src/Symfony/Component/Messenger/Transport/Schedule/Locator/ScheduleLocatorInterface.php:
```diff
@@ -0,0 +1,20 @@
+use Psr\Container\ContainerInterface;
+use Symfony\Component\Messenger\Transport\Schedule\ScheduleConfig;
+
+interface ScheduleLocatorInterface extends ContainerInterface
```
It's more like a `ScheduleConfigLocatorInterface`, right?

upyx commented:
> Now that the implementation is composed of many classes, I'm wondering if we should create a new Scheduler component that would depend on Messenger? We could have shorter namespaces, Lock/Clock/... as "real" dependencies, and better discoverability. That would also avoid "polluting" the Transport sub-namespace as most classes are not about the transport itself. WDYT?

Agreed. At first, I did it like a bridge, but it should be a separate component. What about names? `Schedule` vs `Scheduler`?

https://www.askdifference.com/schedule-vs-scheduler/

However, I'm wondering about development strategy. The first benefits of PR were tiny size and simplicity... but I loved to write code :smile: With the HA feature, it became complicated and... well featured. With a [cron rules parser](https://github.com/dragonmantank/cron-expression) it would cover almost all use cases which I can imagine.

And there is a good full-featured bundle already: https://github.com/Guikingone/SchedulerBundle Do we still need a component like that in-house?

fabpot commented:
> > Now that the implementation is composed of many classes, I'm wondering if we should create a new Scheduler component that would depend on Messenger? We could have shorter namespaces, Lock/Clock/... as "real" dependencies, and better discoverability. That would also avoid "polluting" the Transport sub-namespace as most classes are not about the transport itself. WDYT?
> 
> Agreed. At first, I did it like a bridge, but it should be a separate component. What about names? `Schedule` vs `Scheduler`?
> 
> https://www.askdifference.com/schedule-vs-scheduler/
> 

I'd use Scheduler.

> However, I'm wondering about development strategy. The first benefits of PR were tiny size and simplicity... but I loved to write code smile With the HA feature, it became complicated and... well featured. With a [cron rules parser](https://github.com/dragonmantank/cron-expression) it would cover almost all use cases which I can imagine.
> 
> And there is a good full-featured bundle already: https://github.com/Guikingone/SchedulerBundle Do we still need a component like that in-house?

Having more than one choice is always good. Here, the approach is very different from what already exists, and as such, I think it deserves to be its own component. Maybe we will add more features over time, maybe not. We will see how it goes.

fabpot commented on src/Symfony/Component/Messenger/Transport/Schedule/Trigger/PeriodicalTrigger.php:
```diff
@@ -0,0 +1,110 @@
+        }
+    }
+
+    public static function create(
```
What about making it `__construct` instead?

upyx replied:
It was. It's convenient to have a universal constructor from anything, but it might be difficult to understand how the constructor works and what to expect from it. It is nice to have a way to skip all "the magic" and just set properties.

upyx replied:
Shall we create some factory for triggers with methods like: "once", "hourly", "daily", etc?

upyx commented:
>... We will see how it goes.

:+1:
It's my favorite strategy :smile: 

I've extracted a component. It isn't a standalone component but provides a `Transport` for the Messenger Component. Hence, it configures in the `transport` section of the `messenger`. I doubt that I did the integration with the Messenger Component and the Framework Bundle correctly.

fabpot commented on src/Symfony/Component/Scheduler/Progress/ProgressInterface.php:
```diff
@@ -0,0 +1,25 @@
+
+namespace Symfony\Component\Scheduler\Progress;
+
+interface ProgressInterface
```
I'm wondering is `StateInterface` would be a better name?

upyx replied:
Are you sure? The locking feature doesn't match with a "state". It's more like a point on a schedule or a timeline that can be accessible or not. I've tried: state, position, cursor, progress... Feels like walking in circles. I am not good at it. Thank you for your help :slightly_smiling_face:

fabpot commented on src/Symfony/Component/Scheduler/Trigger/TriggerInterface.php:
```diff
@@ -0,0 +1,17 @@
+
+interface TriggerInterface
+{
+    public function nextTo(\DateTimeImmutable $run): ?\DateTimeImmutable;
```
What about just `next`?

upyx replied:
It was: `getNextRun`, `nextRun`, `next`. Then my colleague asked me why I am set a time here... Now it is `nextTo($run)` that sounds clearer to me. I'd like to name it better if possible.

fabpot replied:
CronExpression names it `getNextRunDate()`, which sounds very descriptive to me.

fabpot reviewed (commented):
I like the new SimpleProgress class that can be decorated. If you think this is the right direction, can you finish the transition? I suppose that SimpleProgress will become the new main Progress class.

upyx commented:
> I like the new SimpleProgress class that can be decorated. If you think this is the right direction, can you finish the transition? I suppose that SimpleProgress will become the new main Progress class.

Yep, decorators are beautiful and shiny. However, they have issues too:

- performance, the `Progress` is used in quite a hot loop,
- order, think how a scheduler should work with a lock but without shared state (in cache)?

As a compromise, I restricted the `LockProgressDecorator` to work over base progress only. Another variant is not using decorators.

fabpot committed:
[Scheduler] Rework the component

fabpot commented:
@upyx I've just pushed 2 commits: one that leverage attributes (like done elsewhere in the framework). The second one renames one class/interface to avoid confusion. Please tell me what you think about these changes. I will have more changes. But let's first wait for your validation here.

fabpot commented:
The last commit introduces `ScheduledMessage` to get rid of the `[Trigger, object]` convention.
It also renames some classes to make things hopefully clearer (some names are probably going to change again, naming is hard).

upyx commented:
@fabpot hi :wave:
Are you still in progress? Could you describe what changes you are making? Maybe I can help?

fabpot commented:
I have difficulty finding time to move forward here, but I've pushed another big chunk of changes.

The biggest change is the addition of a `scheduler.yaml` where one can configure the scheduler. I'm now automatically adding the messenger transports, so all configuration is done in this new file and the fact that we're using Messenger is now an implementation detail (people still need to understand that they need to use messenger:consume to run the scheduler, but that's all). That makes things easier to understand and decouples things a bit more, allowing for greater future configuration extensibility.

The other big change is simplifying the wiring by using the built-in DI features more (the compiler pass, for instance, is no longer needed).

I've also merged all State classes back into one class (as done before) as I agree with you that having decorators that can only be used in one way is suboptimal. I've done it also because I've made Lock and Cache hard dependencies of the component. I think it makes sense to require them, especially as we have good defaults for all platforms out there.

I will try to do the next iteration soon.

fabpot commented:
Here is a configuration example:

```yaml
framework:
    scheduler:
        schedulers:
            default: ~

            dynamic-schedule:
                cache: 'cache.app' # this is the cache pool name
                lock:
                    resource: 'default' # This is the lock factory name
                    ttl: 60
                    auto_release: true
```

upyx commented:
> I have difficulty finding time to move forward here

Let me help you :slightly_smiling_face: 

> The biggest change is the addition of a scheduler.yaml where one can configure the scheduler.

You did it! And I :+1:  it a lot. I'm sure it will pay off in the long run.

> they need to use messenger:consume to run the scheduler

It's possible to configure a different bus that won't conflict with the default one. Then we can create a special command to run schedules.

> I've made Lock and Cache hard dependencies

Are you sure it's a good idea? There are many use cases where mutex or even saving progress between restarts aren't necessary. I'd rather add two implementations of State: with and without dependencies.

> I will try to do the next iteration soon.

:heart:

FYI
Previously, I wondered a lot about separate configurations. The short term benefits of using it as "just transport" are:

1. Simple customization of dispatching.
2. Reusing handlers to run them sync and async by Messenger's routes
3. Simple extention by decorators.

However, hiding the Messenger as an implementation detail has much more profit.

fabpot commented:
I've renamed `ScheduledMessage` to `RecurringMessage` which is more accurate. That made me realize that `OnceTrigger` should not exist (as we cannot guarantee that a single trigger will be triggered -- the consumer might be down just when the trigger would have happened; this is a use case better suited for Messenger itself).

upyx commented:
> OnceTrigger should not exist

So let's delete it. It is not very useful. I've added it to provide some simple example of the `TriggerInterface` implementation. 

> we cannot guarantee that a single trigger will be triggered

Actually can. Trigger implements the logical points where messages are sent, not recieved. Handler recieved the messages that were sent since previous run. The point of previous run can be saved in memory (and lose on restarts) or in some external storage. So all the messages which were scheduled will be handler, sooner or later. For example, 3 messages are handled:

```mermaid
%%{init: { 'gitGraph': {'mainBranchName': 'timeline'}} }%%
gitGraph
  commit id: "previous run" type: HIGHLIGHT
  commit id: "once trigger time"
  commit id: "recurring trigger time1"
  commit id: "recurring trigger time2"
  commit id: "current run" type: HIGHLIGHT
```

To make the messages are recieved as frequently as sent, the Messenger Bus must be processed not less often then "sending" messages are scheduled, and the "previous run" point must not be saved in external storage. It's default behavior of Messenger.

fabpot commented:
I've added a `CronExpressionTrigger` and some shortcuts on `RecurringMessage`. So, you can do:

```php
yield (new RecurringMessage('every-minute', new SomeJob()))->everyMinute();
```

I've also tweaked the transport to use `yield Envelope::wrap($message, [new ScheduledStamp()])` to allow yielding envelopes instead of pure objects:

```php
$e = new Envelope(new SomeJob(), [
    new BusNameStamp('somewhere'),
    new HandlerArgumentsStamp(['foo', 'bar']),
]);

yield new RecurringMessage('every-second', $e, ...);
```

fabpot commented:
@upyx Regarding the `OnceTrigger` trigger, I was more thinking about when the consumer is down when the date of the trigger happens.
So, you run the consumer for a while, you shut it down, the trigger date happens, then you run the consumer again. The message will never be sent, right? We don't "rewind" the clock to catch up with the "lost" messages.

fabpot commented:
Actually, if I understand the code correctly, it DOES catch up with the old messages. but I think it should be optional. If we want to be similar to cron, the default would be to not catch up.

fabpot commented:
I've removed the name for `RecurringMessage`.

upyx commented:
> Actually, if I understand the code correctly, it DOES catch up with the old messages. but I think it should be optional. If we want to be similar to cron, the default would be to not catch up.

Yes, it's optional and disabled by default. The `cache` option can be used for enable it, and it's the only thing a cache used for.

> If we want to be similar to cron...

Do we want?

fabpot commented:
My last iteration:

 * I've renamed `TriggerInterface::nextTo()` tp `TriggerInterface::getNextRunDate()`, which is what CronExpression uses.
 * I've introduced a `Schedule` class that wraps `RecurringMessage` instances to ease usage without the full-stack framework and gives an entry point for more configuration (see next) 
 * I've removed the configuration in `scheduler.yaml`. Now, we can configure everything in a Schedule, which makes it easy to use the Scheduler without Symfony full stack. 
 * `State` instances (now `Checkpoint`) are now created internally to avoid leaking how things work internally
 * Cache is now an optional dependency and not used by default
 * I've added a generic `CallbackTrigger` class
 * I've re-introduced a generic (in Messenger) `RedispatchMessage` class with its handler. I think this is crucial when dealing with "tasks" that take a lot of time (we don't want the scheduler to be waiting and allow the tasks to be handled via another transport/worker).

Here is a usage example that does not use Messenger at all:

```php
<?php

$cache = new FilesystemAdapter('scheduler', 3600, __DIR__.'/scheduler-cache');
$lockFactory = new LockFactory(new InMemoryStore());

$message = new SomeJob('every second simple');
$recurringMessage = (new RecurringMessage($message))
    ->every('1 second')
;
$schedule = (new Schedule())
    ->add($recurringMessage)
    ->lock($lockFactory->createLock('simple-schedule', 600, true))
    ->stateful($cache)
;

$messageGenerator = new MessageGenerator('simple', $schedule);

$handler = new SomeHandler();
while (true) {
    $ran = false;
    foreach ($messageGenerator->getMessages() as $message) {
        $handler($message);
        $ran = true;
    }

    if (!$ran) {
        sleep(1);
    }
}
```

If using Messenger, you can define the transport to avoid the message being handled synchronously when triggered:

```php
#[AsSchedule(name: 'some-schedule')]
class ScheduleProvider implements ScheduleableInterface
{
    public function __construct(
        private CacheInterface $cache,
        private LockFactory $lockFactory,
    ) {
    }

    public function getSchedule(): Schedule
    {
        $message = new SomeJob('every second simple');

        // Message can be a Messenger Envelope if needed
        $envelope = new Envelope($message, [
            new HandlerArgumentsStamp(['foo', 'bar']),
        ]);

        $recurringMessage = (new RecurringMessage($envelope))
            ->every('1 second')
            // send the message to the async transport instead of handling it synchronously
            ->transport('async')
        ;

        return (new Schedule())
            ->add($recurringMessage)
            // avoid having more than one concurrent worker running this schedule
            ->lock($this->lockFactory->createLock('simple-schedule', 600, true))
            // catch up if the consumer if down for some time
            ->stateful($this->cache)
        ;
    }
}
```

jdreesen commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -691,6 +697,10 @@ public function load(array $configs, ContainerBuilder $container)
             $definition->addTag('messenger.message_handler', $tagAttributes);
         });
 
+        $container->registerAttributeForAutoconfiguration(AsSchedule::class, static function (ChildDefinition $definition, AsSchedule $attribute, \ReflectionClass $reflector): void {
```
The last parameter is not necessary, right?

```suggestion
        $container->registerAttributeForAutoconfiguration(AsSchedule::class, static function (ChildDefinition $definition, AsSchedule $attribute): void {
```

nicolas-grekas commented on src/Symfony/Component/Messenger/Message/RedispatchMessage.php:
```diff
@@ -0,0 +1,27 @@
+/**
+ * @internal
+ *
+ * @experimental
```
can be removed

nicolas-grekas commented on src/Symfony/Component/Messenger/Handler/RedispatchMessageHandler.php:
```diff
@@ -0,0 +1,28 @@
+use Symfony\Component\Messenger\Message\RedispatchMessage;
+use Symfony\Component\Messenger\MessageBusInterface;
+
+final class RedispatchMessageHandler
```
let's add a changelog entry about this

fabpot replied:
Done

nicolas-grekas commented on src/Symfony/Component/Messenger/composer.json:
```diff
@@ -21,10 +21,13 @@
         "symfony/dependency-injection": "^5.4|^6.0",
         "symfony/event-dispatcher": "^5.4|^6.0",
         "symfony/http-kernel": "^5.4|^6.0",
+        "symfony/lock": "^5.4|^6.0",
```
messenger needs neither cache nor clock not lock, isn't it?

fabpot replied:
leftover

nicolas-grekas commented on src/Symfony/Component/Scheduler/DependencyInjection/AddScheduleMessengerPass.php:
```diff
@@ -0,0 +1,54 @@
+            $transportDefinition = (new Definition(TransportInterface::class))
+                ->setFactory([new Reference('messenger.transport_factory'), 'createTransport'])
+                ->setArguments(['schedule://'.$name, ['transport_name' => $transportName], new Reference('messenger.default_serializer')])
+                ->addTag('messenger.receiver', [
```
call could be on one line

nicolas-grekas commented on src/Symfony/Component/Scheduler/Generator/MessageGenerator.php:
```diff
@@ -0,0 +1,106 @@
+
+    public function getMessages(): \Generator
+    {
+        if (!$this->waitUntil ||
```
we usually put boolean operator on the start of the next line

fabpot replied:
sure

nicolas-grekas commented on src/Symfony/Component/Scheduler/Messenger/ScheduledStamp.php:
```diff
@@ -0,0 +1,21 @@
+/**
+ * @experimental
+ */
+final class ScheduledStamp implements NonSendableStampInterface
```
maybe we don't need this stamp since we don't use it?

fabpot replied:
@upyx Do you had a use case in mind for that stamp?

upyx replied:
It implements NonSendableStampInterface which a mark for Messnger to handle message but not route to the default queue.

nicolas-grekas commented on src/Symfony/Component/Scheduler/Messenger/SchedulerTransportFactory.php:
```diff
@@ -0,0 +1,63 @@
+        }
+
+        return new SchedulerTransport(
+            new MessageGenerator(
```
should be on one line

nicolas-grekas commented on src/Symfony/Component/Scheduler/README.md:
```diff
@@ -0,0 +1,18 @@
+Scheduler Component
+===================
+
+Provides basic scheduling through the Symfony Messenger.
```
~the~

nicolas-grekas commented on src/Symfony/Component/Scheduler/RecurringMessage.php:
```diff
@@ -0,0 +1,104 @@
+     */
+    public function every(string $expression): self
+    {
+        $this->trigger = PeriodicalTrigger::create(
```
on one line

nicolas-grekas commented on src/Symfony/Component/Scheduler/Schedule.php:
```diff
@@ -0,0 +1,72 @@
+        return $this;
+    }
+
+    public function lock(LockInterface $lock): self
```
`@return $this`

nicolas-grekas commented on src/Symfony/Component/Scheduler/Schedule.php:
```diff
@@ -0,0 +1,72 @@
+        return $this->lock ?? new NoLock();
+    }
+
+    public function stateful(CacheInterface $state): self
```
`@return $this`

nicolas-grekas commented on src/Symfony/Component/Scheduler/Schedule.php:
```diff
@@ -0,0 +1,72 @@
+    private ?LockInterface $lock = null;
+    private ?CacheInterface $state = null;
+
+    public function add(RecurringMessage $message, RecurringMessage ...$messages): self
```
`@return $this`

nicolas-grekas commented on src/Symfony/Component/Scheduler/Schedule.php:
```diff
@@ -0,0 +1,72 @@
+        return $this->messages;
+    }
+
+    public function getSchedule(): self
```
`@return $this`

nicolas-grekas commented on src/Symfony/Component/Scheduler/Tests/Generator/MessageGeneratorTest.php:
```diff
@@ -0,0 +1,173 @@
+        }
+    }
+
+    public function messagesProvider(): \Generator
```
should be static and probably named provideGetMessages

nicolas-grekas commented on src/Symfony/Component/Scheduler/Trigger/ExcludeTimeTrigger.php:
```diff
@@ -0,0 +1,35 @@
+    ) {
+    }
+
+    public function getNextRunDate(\DateTimeImmutable $run): \DateTimeImmutable
```
I guess psalm is right and this should return `?\DateTimeImmutable`

nicolas-grekas commented on src/Symfony/Component/Scheduler/Trigger/PeriodicalTrigger.php:
```diff
@@ -0,0 +1,115 @@
+            return null;
+        }
+
+        $delta = (float) $run->format('U.u') - (float) $this->firstRun->format('U.u');
```
no need for the float casts, php will do it on its own

upyx replied:
It's not for PHP. It's for one who will read the code in the future.

nicolas-grekas commented on src/Symfony/Component/Scheduler/Trigger/PeriodicalTrigger.php:
```diff
@@ -0,0 +1,115 @@
+        /** @var \DateTimeImmutable $nextRun */
+        $nextRun = \DateTimeImmutable::createFromFormat('U.u', $nextRunTimestamp.$this->firstRun->format('.u'));
+
+        if ($this->priorTo <= $nextRun) {
```
let's use a ternary op

nicolas-grekas commented on src/Symfony/Component/Scheduler/composer.json:
```diff
@@ -0,0 +1,40 @@
+{
+    "name": "symfony/scheduler",
+    "type": "library",
+    "description": "Provides basic scheduling through the Symfony Messenger",
```
~the~

nicolas-grekas reviewed (commented):
Nice :)
Here is a CS-level review.
Please figure out a way to squash the PR in two commits if possible :grimacing:

upyx commented:
There are many things were broken...

upyx commented:
@fabpot I've taken a look at your code.

> * I've introduced a `Schedule` class that wraps `RecurringMessage` instances to ease usage without the full-stack framework and gives an entry point for more configuration (see next)

Yes! You got the point :+1:

> * `State` instances (now `Checkpoint`) are now created internally to avoid leaking how things work internally

I like a new name. `CheckpointInterface` is a good extension point that can help to implement different locking/caching strategies. It's an advanced topic. If you'd like to restrict it, the "internal" mark can be used. There is no need to sacrifice a design for hiding how things work. An object's composition should be created in a factory rather than in objects' constructors.

> * Cache is now an optional dependency and not used by default

But Lock? The `Checkpoint` class is the only place where it is optionally needed. I have been using a "message generator transport" in production without locking for two years. It works well.

There was a commit (I can restore it from history) where both cache and lock are optional.

> * I've re-introduced a generic (in Messenger) `RedispatchMessage` class with its handler. I think this is crucial when dealing with "tasks" that take a lot of time (we don't want the scheduler to be waiting and allow the tasks to be handled via another transport/worker).

Have I said that I have a bitter experience with it? My team did it at once. We thought that it was a good idea too. But asynchronous processing of messages doesn't work faster (but a bit slower). So if it is "too long", a queue grows and gets stuck. First, we added more workers which process the queue in parallel. But then we realised that we could just run more workers to process the "schedule" directly!

Generated messages are already "received" from some "chrono queue". To process messages in parallel, you just need to run workers (schedule workers) in parallel.

And "redispatching" is useless, because Messenger's worker dispatches all messages through a bus, and the `NonSendableStampInterface` prevents them from being sent to a queue. We can just remove the mark to have the same effect.

>
> Here is a usage example that does not use Messenger at all:
>
> ```
> <?php
>
> $cache = new FilesystemAdapter('scheduler', 3600, __DIR__.'/scheduler-cache');
> $lockFactory = new LockFactory(new InMemoryStore());
>
> $message = new SomeJob('every second simple');
> $recurringMessage = (new RecurringMessage($message))
>     ->every('1 second')
> ;
> $schedule = (new Schedule())
>     ->add($recurringMessage)
>     ->lock($lockFactory->createLock('simple-schedule', 600, true))
>     ->stateful($cache)
> ;
>
> $messageGenerator = new MessageGenerator('simple', $schedule);
>
> $handler = new SomeHandler();
> while (true) {
>     $ran = false;
>     foreach ($messageGenerator->getMessages() as $message) {
>         $handler($message);
>         $ran = true;
>     }
>
>     if (!$ran) {
>         sleep(1);
>     }
> }
> ```
>

Almost perfect!

The only exception are `RecurringMessage`'s helpers. I'd rather make `RecurringMessage` immutable with static factories, because it's hardly possible to call `everyMinute` and `everyHour` on the same object... and if I do it, how would it work? It confuses.

And about `CronExpressionTrigger`... It is not better than `PeriodicalTrigger`, so helpers should use an internal trigger and not rely on external dependencies.

> If using Messenger, you can define the transport to avoid the message being handled synchronously when triggered:

Yes, it's possible to handle generated messages asynchronously. But it is pointless... Believe me, please.

> I think I'm done for now.

Would you like me to make the changes I described?

upyx commented:
Just for example https://github.com/upyx/symfony/commit/a534c2d26037655fd519cc0d4349816bb664fc8d

fabpot commented:
I've taken everything into account.

lyrixx commented:
Let's go 👍🏼

nicolas-grekas commented on src/Symfony/Component/Messenger/CHANGELOG.md:
```diff
@@ -14,6 +14,7 @@ CHANGELOG
  * Deprecate `StopWorkerOnSigtermSignalListener` in favor of
    `StopWorkerOnSignalsListener` and make it configurable with SIGINT and
    SIGTERM by default
+ * Add `RedispatchMessage` and `RedispatchMessageHandler`
```
should be removed

fabpot replied:
done

nicolas-grekas commented on src/Symfony/Component/Messenger/composer.json:
```diff
@@ -21,6 +21,8 @@
     "require-dev": {
         "psr/cache": "^1.0|^2.0|^3.0",
+        "symfony/cache": "^5.4|^6.0",
+        "symfony/clock": "^6.2",
```
should be removed?

fabpot replied:
done

nicolas-grekas commented on src/Symfony/Component/Scheduler/RecurringMessage.php:
```diff
@@ -0,0 +1,60 @@
+/**
+ * @experimental
+ */
+final class RecurringMessage
```
the class should not be final to me, because it's part of the signature or other classes and doing so kills some extensibiiity

fabpot replied:
done

upyx replied:
More restrictions are good to start. When we will see a reason to remove "filnal", it's easy to remove it. But it's not easy to [add](https://github.com/symfony/symfony/issues/47096#issuecomment-1434333688) it.

I dont's see any usable way to exnend it, so I think it should be final.

fabpot replied:
It's final again :)

nicolas-grekas commented on src/Symfony/Component/Scheduler/RecurringMessage.php:
```diff
@@ -0,0 +1,60 @@
+    /**
+     * Uses a relative date format to define the frequency.
+     *
+     * @see https://www.php.net/manual/en/datetime.formats.relative.php
```
shorter: https://php.net/datetime.formats.relative

fabpot replied:
yep

nicolas-grekas commented on src/Symfony/Component/Scheduler/RecurringMessage.php:
```diff
@@ -0,0 +1,60 @@
+     *
+     * @see https://www.php.net/manual/en/datetime.formats.relative.php
+     */
+    public static function every(string $frequency, object $message, \DateTimeImmutable $from = new \DateTimeImmutable(), ?\DateTimeImmutable $until = new \DateTimeImmutable('3000-01-01')): self
```
static instead of self on all method (matters when making it non-final)

fabpot replied:
oc

stof commented on src/Symfony/Component/Scheduler/RecurringMessage.php:
```diff
@@ -0,0 +1,58 @@
+     *
+     * @see https://php.net/datetime.formats.relative
+     */
+    public static function every(string $frequency, object $message, \DateTimeImmutable $from = new \DateTimeImmutable(), ?\DateTimeImmutable $until = new \DateTimeImmutable('3000-01-01')): static
```
this is broken if calling `ChildRecurringMessage::every()` as it will still return an instance of RecurringMessage, not of ChildRecurringMessage (due to `new self`).

The private constructor does not really allow to instantiate child classes though (and using `new static()` would require making the constructor final (or making the constructor signature part of an implemented interface) to forbid it to change signature.

I suggest keeping `: self` on those 2 methods, and replacing `::trigger` by a public constructor instead (child classes could then extend the constructor safely).

Alternatively, as this object is just a VO combining the TriggerInterface and the message object, we would keep it final (tests could create an actual instance of that VO)

nicolas-grekas commented on src/Symfony/Component/Scheduler/Tests/Messenger/SchedulerTransportTest.php:
```diff
@@ -0,0 +1,68 @@
+
+        $transport->ack(new Envelope(new \stdClass()));
+
+        $this->assertTrue(true); // count coverage
```
should be replaced by expectNotToPerformAssertions at the start of the method

upyx replied:
The `expectNotToPerformAssertions` works different: it's doesn't count code coverage. **This** test covers the logic, so should be counted in coverage.

upyx replied:
I've recheked. Useless test.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Tests/Functional/SchedulerTest.php:
```diff
@@ -0,0 +1,61 @@
+        $this->assertSame([$foo], $fetchMessages(1.0));
+        $this->assertSame([], $fetchMessages(1.0));
+        $this->assertSame([$bar], $fetchMessages(60.0));
+        $this->assertSame([$foo, $bar, $foo, $bar], $fetchMessages(600.0));
```
Tests are green but this result is suspicious, isn't it?

upyx replied:
I'll check. It's funtional, and it's a sequence of "runs".

nicolas-grekas commented on src/Symfony/Component/Scheduler/Generator/MessageGenerator.php:
```diff
@@ -0,0 +1,99 @@
+
+            if ($yield) {
+                yield $message;
+                $this->checkpoint->save($time, $index);
```
I'm wondering here: should we save before or after the yield? Saving after means that if the generator is never resumed (as could happen when an exception occurs), we're going to keep trying the same message. Legit?

upyx replied:
Yes, it should be retried. The Messnger deals with exceptions from here.

upyx commented on src/Symfony/Component/Messenger/Transport/TransportFactory.php:
```diff
@@ -49,6 +49,8 @@ public function createTransport(#[\SensitiveParameter] string $dsn, array $optio
             $packageSuggestion = ' Run "composer require symfony/amazon-sqs-messenger" to install Amazon SQS transport.';
         } elseif (str_starts_with($dsn, 'beanstalkd://')) {
             $packageSuggestion = ' Run "composer require symfony/beanstalkd-messenger" to install Beanstalkd transport.';
+        } elseif (str_starts_with($dsn, 'schedule://')) {
```
Is it still necessary?

fabpot replied:
Probably not

upyx commented on src/Symfony/Component/Scheduler/Messenger/SchedulerTransport.php:
```diff
@@ -0,0 +1,56 @@
+    ) {
+    }
+
+    public function getMessageCount(): int
```
The `MessageCountAwareInterface` need for "fail" queues only. It's not possible to send messages in schedule. Why have you decided to impltment it?

fabpot replied:
It was for `messenger:stats`, but not really useful, I will remove.

upyx commented on src/Symfony/Component/Scheduler/Messenger/SchedulerTransport.php:
```diff
@@ -0,0 +1,56 @@
+    public function get(): iterable
+    {
+        foreach ($this->messageGenerator->getMessages() as $message) {
+            yield Envelope::wrap($message, [new ScheduledStamp()]);
```
Was better. All instances of `ScheduledStamp` are the same, so can be reused.

nicolas-grekas reviewed (approved):
Let's merge and iterate!

fabpot commented:
Thank you @upyx.

upyx commented:
Thank **you** Fabien!

I'm proud to have been working with you :slightly_smiling_face:
